### PR TITLE
WT-12524 Common mechanism for parallel execution.

### PR DESCRIPTION
**Pull request by ershov**

### CODE REVIEW

**Review by ajmorton — COMMENTED** [review]
Some initial questions to kick things off. I still need to work through `s_function` and do a second pass of `common_functions.sh`.

I've touched on this in some comments, but it looks like we've added the following assumptions to our dist scripts:
- Our shell scripts are now re-entrant
- Any argument provided to a file not starting with `-` must be a file name
- The main function of our shell scripts always builds a file list and passes that into subprocesses to be handled
It'd be good to document this somewhere, and it wouldn't hurt to notify the team about this before merging.

File: dist/common_functions.sh [hunk 1]
```diff
@@ -106,3 +122,67 @@ filter_if_fast() {
...
+# Try to use -P to avoid blowing up the system.
+# Usage:
+#   check_xargs_P [NUM_PROCESSES]
+#   xargs $XARGS_P ...
+check_xargs_P() {
+  [[ -z "${XARGS_P:+x}" ]] &&
+    XARGS_P=$(echo echo | xargs -P 1 echo >/dev/null 2>&1 && echo "-P ${1:-20}")
+}
+
+# Check if the script is run in a recursive mode (with files args).
```
[the remark this answers is not in the document]
Comment (ershov) [reply dist/common_functions.sh:140]: Added an explicit statement about it in the comment.

File: dist/common_functions.sh [hunk 2]
```diff
@@ -106,3 +122,67 @@ filter_if_fast() {
...
+is_recursive_mode() {
+  [[ -n "${S_RECURSE:-}" ]] && return 0         # S_RECURSE is set.
+  [[ ${BASH_ARGC:-} -eq 0 ]] && return 1        # No args at all
+  for arg in "${BASH_ARGV[@]:-}"; do
+    [[ "${arg:0:1}" == "-" ]] && return 1;      # Option arg found.
+  done
+  return 0
+}
+
+# Returns true if the script has to find files to process (non-recursive call).
```
[the remark this answers is not in the document]
Comment (ershov) [reply dist/common_functions.sh:151]: Added a comment.

File: dist/common_functions.sh [hunk 3]
```diff
@@ -42,7 +42,7 @@ while len(pids):
 }
 
 # Do normal cleanup on exit or abort.
-setup_trap() { trap 'kill `allchildpids` 2>/dev/null; rm -f $t' 0 1 2 3 13 15; }
+setup_trap() { trap 'kill `allchildpids` 2>/dev/null; rm -f $t*; '"$@" 0 1 2 3 13 15; }
```
Comment (korteland) [inline dist/common_functions.sh:45]: Can this be changed to make sure `$t` is not empty please? I don't want this to turn into `rm -f *` due to a bug or misuse or whatever.

File: dist/common_functions.sh [hunk 4]
```diff
@@ -106,3 +124,68 @@ filter_if_fast() {
...
+#        exit $?
+#    }
+#    ... code for processing files when run as parallel or with file arguments ...
+#
+#  - `is_main_run` checks if the script is run in non-recursive mode (has no file args).
+#  - `find ...` lists files to process.
+#  - `do_in_parallel` restarts the same script in parallel mode.
+#  - `exit $?` exits the script with the return code of `do_in_parallel`.
+#
+# NOTE: `do_in_parallel` can't do `exit` itself since Bash is allowed to execute piped
```
Comment (korteland) [inline dist/common_functions.sh:174]: Would `set -o pipefail` help make this cleaner?

**Review by korteland — COMMENTED** [review]
I like the overall shape of this. I largely agree with Andrew's comments.

One "top-level" suggestion in addition to these two comments: could you please use long arguments whenever reasonable? I've OK to leave existing code alone where it's just been moved/reformatted. But I think most places should be using long flags for readability. It's subjective so I'm happy to chat about this offline.

File: dist/common_functions.sh (hunk 4 above)
Comment (korteland) [reply dist/common_functions.sh:174]: No worries - thanks for the explanation.

**Review by korteland — APPROVED** [review]
LGTM. I would still prefer long flags for subcommands but that's not something the existing code does so I can't reasonably ask you to fix it.

**Comment by ershov** [comment]
@ajmorton @korteland thanks guys for you reviews!

### Change of CLI interface

I like the idea of a more explicit separation of the "main" script and "subprocess" one. It could be done by either having them in separate files or by adding an explicit flag in what mode it should run.

The thing is that the scripts that currently do it, do it in an implicit way by looking at their arguments. And unfortunately, other scripts around use both modes they way it is now. For instance, "clang format" is used as standalone by users, it's run from s_all, and it's called from code generation scripts. Also, turns out developers have their own scripts that rely on the existing CLI interface. If we change it in an incompatible way, it'll make people unhappy.

To summarise it, while I like this idea, I don't know how to do it now. So this PR is rather a refactor of the existing approach with intent to keep the external appearance if everything untouched.

### Re-entrability of scripts

The ones that used parallel execution remain re-entrant just as they were.

I updated the commit message about the newly parallel scripts.

File: dist/common_functions.sh [hunk 5]
```diff
@@ -106,3 +124,68 @@ filter_if_fast() {
     cat
   fi
 }
+
+# Try running xargs with the -P option. If it succeeds, set XARGS_P accordingly.
+# POSIX xargs does not support -P option - it's a GNU extension.
+# -P option limits the maximum number of subprocesses created by xargs.
+# Without -P, xargs will sun only one subprocess at a time.
```
Comment (ajmorton) [inline dist/common_functions.sh:131]:
```suggestion
# Without -P, xargs will run only one subprocess at a time.
```

File: dist/common_functions.sh [hunk 6]
```diff
@@ -106,3 +122,67 @@ filter_if_fast() {
...
+# POSIX xargs does not support -P option - it's a GNU extension.
+# -P option limits the maximum number of subprocesses created by xargs.
+# Without -P, it can create a lot.
+# Try to use -P to avoid blowing up the system.
+# Usage:
+#   check_xargs_P [NUM_PROCESSES]
+#   xargs $XARGS_P ...
+check_xargs_P() {
+  [[ -z "${XARGS_P:+x}" ]] &&
+    XARGS_P=$(echo echo | xargs -P 1 echo >/dev/null 2>&1 && echo "-P ${1:-20}")
```
[the remark this answers is not in the document]
Comment (ajmorton) [reply dist/common_functions.sh:137]: Ah my mistake. I was in Linux mode and forgot `nproc` isn't a cross platform command.
If we never specify `NUM_PROCESSES` I wouldn't mind just setting 20, but I'm happy for you to resolve this comment with the code unchanged

**Review by ajmorton — APPROVED** [review]
LGTM. I have a follow on question about the re-entrant scripts, but I don't want that to block the PR. I'll reach out to you on Slack
